src/Mod/PartDesign/App/FeatureHelix.cpp

```diff
@@ -0,0 +1,490 @@
+            }
+
+            boolOp = refineShapeIfActive(boolOp);
+            Shape.setValue(getSolid(boolOp));
+        }
+        else if(getAddSubType() == FeatureAddSub::Subtractive) {
+
+            BRepAlgoAPI_Cut mkCut(base, result);
+            if (!mkCut.IsDone())
+                return new App::DocumentObjectExecReturn("Subtracting the pipe failed");
+            // we have to get the solids (fuse sometimes creates compounds)
+            TopoDS_Shape boolOp = this->getSolid(mkCut.Shape());
+            // lets check if the result is a solid
+            if (boolOp.IsNull())
+                return new App::DocumentObjectExecReturn("Resulting shape is not a solid");
+
+            int solidCount = countSolids(boolOp);
+            if (solidCount > 1) {
+                return new App::DocumentObjectExecReturn("Pipe: Result has multiple solids. This is not supported at this time.");
+            }
+
+            boolOp = refineShapeIfActive(boolOp);
+            Shape.setValue(getSolid(boolOp));
+        }
+
+        return App::DocumentObject::StdReturn;
+    }
+    catch (Standard_Failure& e) {
+
+        if (std::string(e.GetMessageString()) == "TopoDS::Face")
+            return new App::DocumentObjectExecReturn("Could not create face from sketch.\n"
+                "Intersecting sketch entities in a sketch are not allowed.");
+        else
+            return new App::DocumentObjectExecReturn(e.GetMessageString());
+    }
+    catch (Base::Exception& e) {
+        return new App::DocumentObjectExecReturn(e.what());
+    }
+}
+
+
+
+void Helix::updateAxis(void)
+{
+    App::DocumentObject *pcReferenceAxis = ReferenceAxis.getValue();
+    const std::vector<std::string> &subReferenceAxis = ReferenceAxis.getSubValues();
+    Base::Vector3d base;
+    Base::Vector3d dir;
+    getAxis(pcReferenceAxis, subReferenceAxis, base, dir);
+
+    Base.setValue(base.x,base.y,base.z);
+    Axis.setValue(dir.x,dir.y,dir.z);
+}
+
+
+TopoDS_Shape Helix::generateHelixPath(void)
+{
+    double pitch = Pitch.getValue();
+    double height = Height.getValue();
+    bool leftHanded = LeftHanded.getValue();
+    bool reversed = Reversed.getValue();
+    double angle = Angle.getValue();
+    if (angle < Precision::Confusion() && angle > -Precision::Confusion())
+        angle = 0.0;
+
+    // get revolve axis
+    Base::Vector3d b = Base.getValue();
+    gp_Pnt pnt(b.x,b.y,b.z);
+    Base::Vector3d v = Axis.getValue();
+    gp_Dir dir(v.x,v.y,v.z);
+
+    Base::Vector3d normal = getProfileNormal();
+    Base::Vector3d start = v.Cross(normal);  // pointing towards the desired helix start point.
+    gp_Dir dir_start(start.x, start.y, start.z);
+
+    // Determine radius as the minimum distance between sketchshape and axis.
+    // also find out in what quadrant relative to the axis the profile is located.
+    double radius = 1e99;
+    bool turned = false;
+    double startOffset = 1e99;
```

Comment:
I have a general distrust of magic numbers: you might consider:
```suggestion
    double radius = std::numeric_limits<double>::max();
    bool turned = false;
    double startOffset = std::numeric_limits<double>::max();
```